unstructured/utils.py

```diff
@@ -280,3 +281,280 @@ def scarf_analytics():
+    else:
+        n = len(string_B.split()) - 1
+        string_A, string_B = string_B, string_A
+    n_str = str(n)
+    ngram_percentage = 0
+    while not ngram_percentage:
+        ngram_percentage, shared_ngrams = calculate_shared_ngram_percentage(string_A, string_B, n)
+        if n == 0:
+            break
+        else:
+            n -= 1
+    return round(ngram_percentage, 2), shared_ngrams, n_str
+
+
+def is_parent_box(
+    parent_target,
+    child_target,
+    add=0,
+) -> bool:
+    """True if the child_target bounding box is nested in the parent_target.
+    Box format: [x_bottom_left, y_bottom_left, x_top_right, y_top_right].
+    The parameter 'add' is the pixel error tolerance for extra pixels outside the parent region"""
+
+    if len(parent_target) != 4:
+        return False
+
+    if add and len(parent_target) == 4:
+        parent_target = list(parent_target)
+        parent_target[0] -= add
+        parent_target[1] -= add
+        parent_target[2] += add
+        parent_target[3] += add
+
+    if len(child_target) == 4:
+        if (child_target[0] >= parent_target[0] and child_target[1] >= parent_target[1]) and (
+            child_target[2] <= parent_target[2] and child_target[3] <= parent_target[3]
+        ):
+            return True
+    elif len(child_target) == 2:  # Needed for polygon regions, this might need revision
+        if (
+            parent_target[0] <= child_target[0] <= parent_target[2]
+            and parent_target[1] <= child_target[1] <= parent_target[3]
+        ):
+            return True
+
+    return False
+
+
+def calculate_overlap_percentage(box1, box2, intersection_ratio_method="total"):
+    """Box format: [x_bottom_left, y_bottom_left, x_top_right, y_top_right]"""
+    x1, y1 = box1[0]
+    x2, y2 = box1[2]
+    x3, y3 = box2[0]
+    x4, y4 = box2[2]
+    area_box1 = (x2 - x1) * (y2 - y1)
+    area_box2 = (x4 - x3) * (y4 - y3)
+    x_intersection1 = max(x1, x3)
+    y_intersection1 = max(y1, y3)
+    x_intersection2 = min(x2, x4)
+    y_intersection2 = min(y2, y4)
+    intersection_area = max(0, x_intersection2 - x_intersection1) * max(
+        0,
+        y_intersection2 - y_intersection1,
+    )
+    max_area = max(area_box1, area_box2)
+    min_area = min(area_box1, area_box2)
+    total_area = area_box1 + area_box2
+    overlap_percentage = 0
+
+    if intersection_ratio_method == "parent":
+        if max_area == 0:
+            return 0
+        overlap_percentage = (intersection_area / max_area) * 100
+
+    elif intersection_ratio_method == "partial":
+        if min_area == 0:
+            return 0
+        overlap_percentage = (intersection_area / min_area) * 100
+
+    else:
+        if (area_box1 + area_box2) == 0:
+            return 0
+
+        overlap_percentage = (intersection_area / (area_box1 + area_box2 - intersection_area)) * 100
+
+    return round(overlap_percentage, 2), max_area, min_area, total_area
+
+
+def catch_overlapping_bboxes(
+    elements,
+) -> bool:
```

Comment:
This function is over 150 lines long, which is way over the rule of thumb limit of 5-20 lines. Breaking it up into logical subtasks that get their own functions would really increase the readability.

Reply:
I renamed `catch_overlapping_bboxes` to `catch_overlapping_and_nested_bboxes`and create two more methods to break it up: `identify_overlapping_or_nesting_case` and `identify_overlapping_case`. A short description can be found in the [PR description](https://github.com/Unstructured-IO/unstructured/pull/1803#issue-1952887604). Is it better? I can try to split it further but most of the conditions to classify the overlapping-case shall be in one method, making it large.